File: web_client/src/utils/nas.js
import { useState } from 'react';
import { apiFetch } from './api';
import { usePolling } from './usePolling';

// Two sources of truth for "can this archived item be played", used in order of freshness:
//
//  1. `availableNodeIds` — polled from /api/nas/availability, so it stays current while the
//     page is open. This is the one that matters: /api/library is fetched once per page load
//     (useLibraryActions), so without a poll a tab left open while a node went down would keep
//     showing its items as playable.
//  2. `item.nas_available` — stamped onto the row by /api/library (server/src/nasSource.js).
//     Correct as of that fetch, and the fallback for the first render before the poll answers,
//     or against a server too old to have the endpoint.
//
// Both are checked so that a missing signal never reads as unavailable: a locally-stored row
// has no nas_node_id at all, and an older server sends neither field.
export const isNasOffline = (item, availableNodeIds = null) => {
    if (!item) return false;
    if (availableNodeIds && item.nas_node_id) return !availableNodeIds.has(item.nas_node_id);
    return item.nas_available === false;
};

export const nasOfflineMessage = (item) => {
    const label = (item?.title || item?.filename || 'This item').trim();
    const node = item?.nas_node_id ? `NAS node "${item.nas_node_id}"` : 'its NAS node';
    return `${label} is stored on ${node}, which is currently offline.\n\n`
        + `Bring the node back online to watch or restore it.`;
};

const POLL_MS = 10_000;
const TRANSFER_POLL_MS = 1_500;

/**
 * Returns a Set of reachable NAS node ids, or null while unknown — the caller must treat null
 * as "fall back to the row's own flag", never as "nothing is available".
 *
 * Null is also the resting state on a server without the endpoint: a 404 throws here, which
 * usePolling counts as a failed tick and backs off, and the set stays null so every archived
 * item keeps whatever /api/library said.
 */
export const useNasAvailability = (serverUrl, token) => {
    const [availableNodeIds, setAvailableNodeIds] = useState(null);

    usePolling(async () => {
        if (!token) return;
        const res = await apiFetch(serverUrl, '/api/nas/availability', token);
        if (!res.ok) throw new Error(`nas availability failed: ${res.status}`);
        const data = await res.json();
        setAvailableNodeIds(new Set(data.available || []));
    }, POLL_MS, [serverUrl, token]);

    return availableNodeIds;
};

/**
 * Live progress for archive/restore transfers currently in flight, keyed by filename.
 *
 * The nas-action request itself blocks until the whole transfer finishes, so it can never
 * report progress — this polls /api/nas/jobs (the same node-reported byte counters the admin
 * dashboard already shows) in parallel, only while `activeFilenames` is non-empty. Matching by
 * filename alone, same as the node's own job tracking, since nothing ties a job back to a
 * specific /api/media/nas-action call.
 *
 * `activeFilenames` only needs to change identity for the polling loop itself to start/stop —
 * usePolling reads the latest closure on every tick regardless of its deps array, so a filename
 * being added or removed mid-poll is picked up on the very next tick without restarting it.
 */
export const useNasTransferProgress = (serverUrl, token, activeFilenames) => {
    const [jobsByFilename, setJobsByFilename] = useState({});
    const hasActive = activeFilenames.length > 0;

    usePolling(async () => {
        if (!token || !hasActive) { setJobsByFilename({}); return; }
        const res = await apiFetch(serverUrl, '/api/nas/jobs', token);
        if (!res.ok) throw new Error(`nas jobs failed: ${res.status}`);
        const data = await res.json();
        const next = {};
        for (const job of data.jobs || []) next[job.filename] = job;
        setJobsByFilename(next);
    }, TRANSFER_POLL_MS, [serverUrl, token, hasActive]);

    return jobsByFilename;
};

File: StreamPiTV/app/src/main/java/com/example/streampitv/util/Nas.kt
package com.example.streampitv.util

import com.example.streampitv.data.VideoItem

/**
 * Whether an archived item's storage node is down, preferring live data over the stamp.
 *
 * Two sources answer this, in order of freshness:
 *  1. [availableNodeIds], polled from /api/nas/availability — current while the screen is up.
 *  2. [VideoItem.isNasOffline], the `nas_available` stamped on the row by /api/library — only
 *     accurate as of that fetch, and the fallback before the first poll returns or against a
 *     server without the endpoint.
 *
 * The stamp stays the single definition of the fallback, so there is one place this can be
 * wrong rather than two. A null set means "not known yet" and must never read as unavailable —
 * that would make every archived item look broken for the first poll interval.
 */
fun isNasOffline(item: VideoItem, availableNodeIds: Set<String>?): Boolean {
    val node = item.nasNodeId
    if (availableNodeIds != null && node != null) return node !in availableNodeIds
    return item.isNasOffline
}

/**
 * What to tell the viewer when an archived item's storage node is down.
 *
 * Shared by Home, Series detail and the player so the three can't drift, and worded to name the
 * node because that is the only actionable part: the file itself is fine, and both streaming and
 * restoring go through the same node, so there is no alternative route to offer.
 */
fun nasOfflineNotice(item: VideoItem): String {
    // Titles come straight from the media row and some carry a trailing space, which otherwise
    // shows up as a double space mid-sentence.
    val label = (item.title ?: item.filename ?: "This item").trim()
    val node = item.nasNodeId?.let { "NAS node \"$it\"" } ?: "its NAS node"
    return "$label is on $node, which is offline."
}
